src/services/bunker.py

```diff
@@ -1,15 +1,413 @@
+        for current frame for Lido validators
+        """
+        current_ref_epoch = EpochNumber(blockstamp.slot_number // self.c_conf.slots_per_epoch)
+        last_report_blockstamp = get_first_non_missed_slot(
+            self.w3.cc, self.last_report_ref_slot, self.c_conf.slots_per_epoch * self.f_conf.epochs_per_frame
+        )
+        last_member_report_ref_epoch = EpochNumber(last_report_blockstamp.slot_number // self.c_conf.slots_per_epoch)
+
+        total_ref_effective_balance = self._calculate_total_effective_balance(self.all_validators)
+        total_ref_lido_effective_balance = self._calculate_total_effective_balance(
+            [v.validator for v in self.lido_validators]
+        )
+
+        if last_member_report_ref_epoch <= self.f_conf.initial_epoch:
+            # If no one report was made - we should use initial epoch as last completed epoch
+            # Because in this case 'cl_rebase' will contain all rewards from initial epoch
+            last_completed_epoch = self.f_conf.initial_epoch
+            total_last_effective_balance = total_ref_effective_balance
+            total_last_lido_effective_balance = total_ref_lido_effective_balance
+        else:
+            last_completed_epoch = last_member_report_ref_epoch
+            last_all_validators = self.w3.cc.get_validators(last_report_blockstamp.state_root)
+            last_lido_validators = self.w3.lido_validators.filter_lido_validators(self.lido_keys, last_all_validators)
+            total_last_effective_balance = self._calculate_total_effective_balance(last_all_validators)
+            total_last_lido_effective_balance = self._calculate_total_effective_balance(
+                [v.validator for v in last_lido_validators]
+            )
+
+        mean_total_effective_balance = (total_ref_effective_balance + total_last_effective_balance) // 2
+        mean_total_lido_effective_balance = (
+            (total_ref_lido_effective_balance + total_last_lido_effective_balance) // 2
+        )
+
+        epochs_passed = current_ref_epoch - last_completed_epoch
+
+        normal_cl_rebase = self._calculate_normal_cl_rebase(
+            epochs_passed, mean_total_lido_effective_balance, mean_total_effective_balance
+        )
+
+        logger.info({"msg": f"Normal CL rebase: {normal_cl_rebase} Gwei"})
+        return Gwei(normal_cl_rebase)
+
+    def _calculate_nearest_and_far_cl_rebase(self, blockstamp: BlockStamp) -> tuple[Gwei, Gwei]:
+        """
+        Calculate CL rebase from nearest and far epochs to ref epoch given the changes in withdrawal vault
+        """
+        logger.info({"msg": "Calculating nearest and far CL rebase"})
+        nearest_slot = (blockstamp.slot_number - NEAREST_EPOCH_DISTANCE * self.c_conf.slots_per_epoch)
+        far_slot = (blockstamp.slot_number - FAR_EPOCH_DISTANCE * self.c_conf.slots_per_epoch)
+
+        lookup_max_deep = self.c_conf.slots_per_epoch * self.f_conf.epochs_per_frame
```

Comment:
I wonder if a whole day can be missed? 🤔 Let's say in the case of an inactive leak and we are in a separate smaller validator group

Reply:
I think it's okay to make the assumption that this will never happen, but we could, for example, put it in a separate method and add a description that we thought about it and decided that this level of paranoia is enough